### PR TITLE
Added `Convert to Operator` Groovy inspection

### PR DESCRIPTION
**Pull request by paplorinc**
Fixes https://youtrack.jetbrains.com/issue/IDEA-95818
```
a.bitwiseNegate()       → ~a       
a.negative()            → -a       
a.positive()            → +a       
a.call()                → a()      
a.next()                → ++a      
a.previous()            → --a      

a.asBoolean()           → !!a      
!a.asBoolean()          → !a       

if (a.asBoolean())      → if (a)
if (!a.asBoolean())     → if (!a)

a.minus(b)              → a - b    
a.plus(b)               → a + b    
a.power(b)              → a**b     
a.div(b)                → a / b    
a.mod(b)                → a % b    
a.or(b)                 → a | b    
a.and(b)                → a & b    
a.xor(b)                → a ^ b    
a.leftShift(b)          → a << b   
a.rightShift(b)         → a >> b   
a.rightShiftUnsigned(b) → a >>> b  

a.asType(b)             → a as b   
a.multiply(b)           → a * b    
b.isCase(a)             → a in b   

a.equals(b)             → a == b   
!a.equals(b)            → a != b   

a.compareTo(b)          → a <=> b  
a.compareTo(b) < 0      → a < b    
a.compareTo(b) <= 0     → a <= b   
a.compareTo(b) == 0     → a == b   
a.compareTo(b) != 0     → a != b   
a.compareTo(b) >= 0     → a >= b   
a.compareTo(b) > 0      → a > b    

a.getAt(b)              → a[b]     
a.putAt(b, c)           → a[b] = c 
```

### CODE REVIEW

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/GroovyInspectionBundle.properties [hunk 1]
```diff
@@ -47,6 +47,10 @@ groovy.probable.bugs=Probable bugs
 equals.between.inconvertible.types='equals()' between objects of inconvertible types
 equals.between.inconvertible.types.tooltip=<code>{0}</code> between objects of inconvertible types ''{1}'' and ''{2}''
 
+convert.to.operator=Change ''{0}'' to ''{1}''
+convert.to.operator.double.negation.option=Use double negation (i.e. !!)
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/GroovyInspectionBundle.properties]: it's configurable via these two options

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/ChangeToOperatorInspection.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    return new BaseInspectionVisitor() {
+      @Override
+      public void visitMethodCallExpression(GrMethodCallExpression methodCallExpression) {
+        super.visitMethodCallExpression(methodCallExpression);
+        processMethodCall(methodCallExpression);
+      }
+
+      private void processMethodCall(GrMethodCallExpression methodCallExpression) {
+        try {
+          PsiMethod method = methodCallExpression.resolveMethod();
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/ChangeToOperatorInspection.java]: `reolveMethod` isn't working for complex scenarios, though (e.g. `[1,2,3].isCase(0)`)

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/ChangeToOperatorInspection.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+          if (!isValid(method)) return;
+
+          String methodName = method.getName();
+          Transformation transformation = TRANSFORMATIONS.get(methodName);
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/ChangeToOperatorInspection.java]: get the strategy based on the method name

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  public int getParameterCount() {
+    return arguments.length;
+  }
+
+  @Nullable
+  public IElementType getComparison() {
+    return comparison;
+  }
+
+  private static final class Builder {
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java]: build `MethodCallData` based on the parent `PsiElement`

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    private static void validateNotNegated(MethodCallData result) {
+      if (result.isNegated)
+        throw new UnsupportedOperationException(format("Negated comparison not supported yet: '%s'!", result.backingElement.getText()));
+    }
+
+    private static void validateRightOperand(MethodCallData result, GrBinaryExpression binaryExpression) {
+      GrExpression rightOperand = binaryExpression.getRightOperand();
+      if ((rightOperand == null) || !"0".equals(rightOperand.getText()))
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java]: e.g. `compareTo(...) < 1` isn't supported

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/IsCaseTransformation.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import org.jetbrains.plugins.groovy.lang.lexer.GroovyTokenTypes;
+
+class IsCaseTransformation extends BinaryTransformation {
+  public IsCaseTransformation() {
+    super(GroovyTokenTypes.kIN);
+  }
+
+  @Override
+  protected String getLhs(MethodCallData methodInfo) {
+    return super.getRhs(methodInfo);
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/IsCaseTransformation.java]: isCase is inverted, i.e. `a.isCase(b)` is `b in a`

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/PutAtTransformation.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  private final Transformation myGetAt;
+
+  public PutAtTransformation(Transformation getAtTransformtion) {
+    super(null, 2);
+    myGetAt = getAtTransformtion;
+  }
+
+  @Override
+  public String getReplacement(MethodCallData methodInfo, OptionsData optionsData) {
+    return format("%s = %s", myGetAt.getReplacement(methodInfo, optionsData), methodInfo.getArgument(1));
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/PutAtTransformation.java]: reusing the `getAtTransformtion` for the first part

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/PutAtTransformation.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.jetbrains.plugins.groovy.codeInspection.changeToOperator.transformations;
+
+import org.jetbrains.plugins.groovy.codeInspection.changeToOperator.data.MethodCallData;
+import org.jetbrains.plugins.groovy.codeInspection.changeToOperator.data.OptionsData;
+
+import static java.lang.String.format;
+
+class PutAtTransformation extends Transformation {
+  private final Transformation myGetAt;
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/PutAtTransformation.java]: should probably be renamed to `getAtTransformtion` also

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/Transformation.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+  @Nullable
+  public ReplacementData transform(GrMethodCallExpression callExpression, OptionsData optionsData) {
+    try {
+      GrExpression element = getExpandedElement(callExpression);
+      MethodCallData methodInfo = getValidMethodInfo(element);
+      String replacement = getReplacement(methodInfo, optionsData);
+      return new ReplacementData(element, replacement);
+    }
+    catch (UnsupportedOperationException e) {
```
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/transformations/Transformation.java]: it's caught anyway outside ...

**Comment by paplorinc** [comment]
@dovchinnikov, @maxmedvedev

**Comment by paplorinc** [comment]
dovchinnikov, @maxmedvedev

**Comment by paplorinc** [comment]
@dovchinnikov, @maxmedvedev, @ignatov ?

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @Nullable
+    private static PsiElement handleMethodCall(MethodCallData result, PsiElement element) {
+      if (!(element instanceof GrMethodCall)) return null;
+
+      GrMethodCall methodCall = (GrMethodCall)element;
+      GrReferenceExpression invokedExpression = (GrReferenceExpression)methodCall.getInvokedExpression();
+
+      result.base = addParenthesesIfNeeded(invokedExpression.getQualifierExpression());
```
Comment (dovchinnikov) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java]: What if qualifier expression is null?
for example `bar(foo)` is method call, invoked expression is `bar` and it has no qualifier.

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  @NotNull
+  public String getBase() {
+    return base.getText();
+  }
+
+  @NotNull
+  public String getArgument(int i) {
+    return arguments[i].getText();
```
Comment (dovchinnikov) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java]: I get *ArrayIndexOutOfBoundsException* of errors, when I try to type `a.isCase(<caret>)` in editor

**Comment by dovchinnikov** [comment]
I'm sorry for long responses btw.

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java (hunk 11 above)
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java]: Good point, thanks, fixed!

File: plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java (hunk 10 above)
Comment (paplorinc) [inline plugins/groovy/groovy-psi/src/org/jetbrains/plugins/groovy/codeInspection/changeToOperator/data/MethodCallData.java]: Interesting, thanks, will push a fix shortly! :)

**Comment by paplorinc** [comment]
Added a few more error checks for places which I thought were already covered, I hope all `NPE`s and `AIOOBE`s are eliminated now :)

What should I do about the untestable cases (they're working from the IDE, not from JUnit)?

File: plugins/groovy/test/org/jetbrains/plugins/groovy/inspections/GrChangeToOperatorTest.groovy [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+  void testNegatableUnaryExpression() {
+    assertValid(/a.asBoolean()/, /!!a/)
+    assertValid(/!a.asBoolean()/, /!a/)
+
+    assertValid(/if (${_}a.asBoolean()${_});/, /a/)
+    assertValid(/if (${_}!a.asBoolean()${_});/, /!a/)
+  }
+
+  // TODO
```
Comment (paplorinc) [inline plugins/groovy/test/org/jetbrains/plugins/groovy/inspections/GrChangeToOperatorTest.groovy]: these are working correctly from the IDE

Comment (paplorinc) [inline plugins/groovy/test/org/jetbrains/plugins/groovy/inspections/GrChangeToOperatorTest.groovy]: these are working correctly from the IDE

Comment (dovchinnikov) [inline plugins/groovy/test/org/jetbrains/plugins/groovy/inspections/GrChangeToOperatorTest.groovy]: Try to debug, locate the place with execution path different from the IDE one.

Comment (paplorinc) [inline plugins/groovy/test/org/jetbrains/plugins/groovy/inspections/GrChangeToOperatorTest.groovy]: @dovchinnikov, as I stated in a previous comment also
> `methodCallExpression.reolveMethod` isn't working for complex scenarios (e.g. [1,2,3].isCase(0))

I guess a separate issue should be opened for this.

**Comment by dovchinnikov** [comment]
@paplorinc JFYI https://github.com/JetBrains/intellij-community/commit/5153f72ee6c3e88df655e3dabce1797acfd45375

**Pull request merged**